controller/update.go

```diff
@@ -0,0 +1,255 @@
+	}
+	closer = make(<-chan struct{})
+	taskObject, err = c.WaitForTask(taskObject.ID, closer)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	if task.HasFailedStatus(taskObject) {
+		fail <- maskAny(fmt.Errorf(taskObject.Error))
+		return
+	}
+}
+
+// UpdateOptions represents the options defining the strategy of an update
+// process. Lets have a look at how the update process of 3 group slices would
+// look like using the given options.
+//
+//     TODO I am not that happy with this visualization. Improving it? Removing it?
+//
+//     MaxGrowth    1
+//     MinAlive     2
+//     ReadySecs    30
+//
+//     @1 (running)  ->  @1 (stopped/destroyed)
+//     @2 (running)  ->  @2 (running)            ->  @2 (stopped/destroyed)
+//     @3 (running)  ->  @3 (running)            ->  @3 (running)            ->  @3 (stopped/destroyed)
+//                   ->  @1 (submitted/running)  ->  @1 (running)            ->  @1 (running)            ->  @1 (running)
+//                                               ->  @2 (submitted/running)  ->  @2 (running)            ->  @2 (running)
+//                                                                           ->  @3 (submitted/running)  ->  @3 (running)
+//
+type UpdateOptions struct {
+	// MaxGrowth represents the number of groups allowed to be added at a given
+	// time. No more than MaxGrowth groups will be added at the same time during
+	// the update process.
+	MaxGrowth int
+
+	// MinAlive represents the number of groups required to stay healthy during
+	// the update process. No more than MinAlive groups will be removed at the
+	// same time during the update process.
+	MinAlive int
+
+	// ReadySecs represents the number of seconds required to wait before ending
+	// the update process of one group and starting the update process of another
+	// group. This is basically a cool down where the update process sleeps
+	// before updating the next group.
+	ReadySecs int
+}
+
+func (c controller) UpdateWithStrategy(req Request, opts UpdateOptions) error {
+	done := make(chan struct{}, 1)
+	fail := make(chan error, 1)
+	addQueue, removeQueue, err := c.createJobQueues(req, opts)
+	if err != nil {
+		return maskAny(err)
+	}
+	numTotal := len(req.SliceIDs)
+
+	for {
+		select {
+		case jobDesc := <-addQueue:
+			go func(jobDesc updateJobDesc) {
+				for {
+					ok, err := c.isGroupAdditionAllowed(req, numTotal, opts)
+					if err != nil {
+						fail <- maskAny(err)
+						return
+					}
+					if !ok {
+						time.Sleep(c.WaitSleep)
+					}
+					break
+				}
+
+				c.addWorker(jobDesc, removeQueue, fail)
+			}(jobDesc)
+		case jobDesc := <-removeQueue:
+			go func(jobDesc updateJobDesc) {
+				for {
+					ok, err := c.isGroupRemovalAllowed(req, opts)
+					if err != nil {
+						fail <- maskAny(err)
+						return
+					}
+					if !ok {
+						time.Sleep(c.WaitSleep)
+					}
+					break
+				}
+
+				c.removeWorker(jobDesc, fail)
+			}(jobDesc)
+		case err := <-fail:
+			close(done)
+			close(addQueue)
```

Comment:
> Which writer should close the channel in case there are multiple writers using it?

I would say no one. A signal (cancel chan?) should tell those writes to stop and when all are stopped, some upper layer can close all channels. The upper layer would still be this function here, but we need to ensure that the workers are stopped